interfaces/builtin/dbus_app.go

```diff
@@ -0,0 +1,280 @@
+		}
+		busNamesList, ok := attribs[bus].([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("bus attribute is not a list")
+		}
+
+		for _, item := range busNamesList {
+			busName, ok := item.(string)
+			if !ok {
+				return nil, fmt.Errorf("session element is not a string")
+			}
+
+			// https://dbus.freedesktop.org/doc/dbus-specification.html#message-protocol-names
+			if !ok || len(busName) == 0 {
+				return nil, fmt.Errorf("bus name must be set")
+			} else if len(busName) > 255 {
+				return nil, fmt.Errorf("bus name is too long (must be <= 255)")
+			}
+
+			validBusName := regexp.MustCompile("^[a-zA-Z][a-zA-Z0-9_-]*(\\.[a-zA-Z0-9_-]+)+$")
+			if !validBusName.MatchString(busName) {
+				return nil, fmt.Errorf("invalid bus name: %q", busName)
+			}
+
+			busNames[bus] = append(busNames[bus], busName)
+		}
+	}
+
+	return busNames, nil
+}
+
+// verify that name for bus is in list
+func (iface *DbusAppInterface) verifyNameInAttributes(bus string, name string, attribs map[string]interface{}) bool {
+	otherBusNames, err := iface.getBusNames(attribs)
+	if err != nil {
+		return false
+	}
+
+	if otherNames, ok := otherBusNames[bus]; ok {
+		for _, otherName := range otherNames {
+			if name == otherName {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// Determine AppArmor dbus abstraction to use based on bus
+func getAppArmorAbstraction(bus string) (string, error) {
+	var abstraction string
+	if bus == "system" {
+		abstraction = "dbus-strict"
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%s'", bus)
+	}
+	return abstraction, nil
+}
+
+// Calculate individual snippet policy based on bus and name
+func getAppArmorIndividualSnippet(policy []byte, bus string, name string) []byte {
+	old := []byte("###DBUS_BIND_BUS###")
+	new := []byte(bus)
+	snippet := bytes.Replace(policy, old, new, -1)
+
+	old = []byte("###DBUS_BIND_NAME###")
+	new = []byte(name)
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path
+	dot_re := regexp.MustCompile("\\.")
+	var path_buf bytes.Buffer
+	path_buf.WriteString(`"/`)
+	path_buf.WriteString(dot_re.ReplaceAllString(name, "/"))
+	path_buf.WriteString(`{,/**}"`)
+
+	old = []byte("###DBUS_BIND_PATH###")
+	new = path_buf.Bytes()
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	return snippet
+}
+
+func (iface *DbusAppInterface) SanitizePlug(plug *interfaces.Plug) error {
+	if iface.Name() != plug.Interface {
+		panic(fmt.Sprintf("plug is not of interface %q", iface))
+	}
+
+	_, err := iface.getBusNames(plug.Attrs)
+	return err
+}
+
+func (iface *DbusAppInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	_, err := iface.getBusNames(slot.Attrs)
+	return err
+}
+
+func (iface *DbusAppInterface) AutoConnect() bool {
+	return false
```

Comment:
Since we are implementing this in the permanent slot side it is auto-generated. There is no plugs side atm. I supposed it would be clearer if I just removed this.